client/src/client.rs

```diff
@@ -170,936 +88,196 @@ impl Client {
-                if let Some(ip) = state.virt_ips.remove(&(slot, addr)) {
-                    state.virt_nodes.remove(&ip);
-                }
+                Some(addr) => self.state.write().await.public_addr = Some(addr),
+                None => self.state.write().await.public_addr = None,
             }
         }
 
-        let removed = state.sessions.remove(&addr);
-        drop(state);
-
-        if let Some(session) = removed {
-            let mut session_state = session.state.write().await;
-            *session_state = None;
-        }
-    }
-
-    async fn register_endpoints(
-        &self,
-        addr: SocketAddr,
-        session_id: SessionId,
-        endpoints: Vec<proto::Endpoint>,
-    ) -> anyhow::Result<Vec<proto::Endpoint>> {
-        let id = self.id();
-        log::info!("[{}] registering endpoints", id);
-
-        let response = self
-            .request::<proto::response::Register>(
-                proto::request::Register { endpoints }.into(),
-                session_id.to_vec(),
-                REGISTER_REQUEST_TIMEOUT,
-                addr,
-            )
-            .await?
-            .packet;
-
-        log::info!("[{}] registration finished", id);
-
-        Ok(response.endpoints)
-    }
-
-    async fn find_node(
-        &self,
-        addr: SocketAddr,
-        session_id: SessionId,
-        node_id: NodeId,
-    ) -> anyhow::Result<proto::response::Node> {
-        let packet = proto::request::Node {
-            node_id: node_id.into_array().to_vec(),
-            public_key: true,
-        };
-        self.find_node_by(addr, session_id, packet).await
+        log::debug!("[{}] started", self.node_id());
+        Ok(())
     }
 
-    async fn find_slot(
-        &self,
-        addr: SocketAddr,
-        session_id: SessionId,
-        slot: SlotId,
-    ) -> anyhow::Result<proto::response::Node> {
-        let packet = proto::request::Slot {
-            slot,
-            public_key: true,
-        };
-        self.find_node_by(addr, session_id, packet).await
+    pub async fn forward(&self, node_id: NodeId) -> anyhow::Result<ForwardSender> {
+        // Establish session here, if it didn't existed.
+        let session = self.sessions.optimal_session(node_id).await?;
+        self.sessions.forward(session, node_id).await
     }
 
-    async fn find_node_by(
-        &self,
-        addr: SocketAddr,
-        session_id: SessionId,
-        packet: impl Into<proto::Request>,
-    ) -> anyhow::Result<proto::response::Node> {
-        let response = self
-            .request::<proto::response::Node>(
-                packet.into(),
-                session_id.to_vec(),
-                DEFAULT_REQUEST_TIMEOUT,
-                addr,
-            )
-            .await?
-            .packet;
+    pub async fn forward_unreliable(&self, node_id: NodeId) -> anyhow::Result<ForwardSender> {
+        log::trace!("{} forward_unreliable to {}", self.config.node_id, node_id);
```

Comment:
> why did this log update to trace? i think debug was fine

This stops being fine, when we look at yagna logs. We can change this later, but I prefer to look in bigger picture, because yagna logs are sometimes annoying